src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -439,6 +440,11 @@ private BoundLambda ReallyBind(NamedTypeSymbol delegateType)
                 }
             }
 
+            // when binding for real (not for return inference), there is still
+            // a good chance that we could reuse a body of a lambda previously bound for 
+            // return type inference.
+            var cacheKey = ReturnInferenceCacheKey.Create(delegateType, IsAsync);
```

Comment:
>cacheKey [](start = 16, length = 8)

This cache key is not used as a cache key. In fact, it really isn't used, except to get things that are available on the delegate type.

Reply:
My point is, please avoid creating the cache key, and instead directly compute the parameter type array and parameter ref kind array.

Reply:
➡️ Reverted the key change; can evaluate separately.